canal/sync.go

```diff
@@ -122,42 +114,38 @@ func (c *Canal) runSyncBinlog() error {
 			if e.GSet != nil {
 				c.master.UpdateGTIDSet(e.GSet)
 			}
-			var (
-				mb    [][]byte
-				db    []byte
-				table []byte
-			)
-			regexps := []regexp.Regexp{*expCreateTable, *expAlterTable, *expRenameTable, *expDropTable, *expTruncateTable}
-			for _, reg := range regexps {
-				mb = reg.FindSubmatch(e.Query)
-				if len(mb) != 0 {
-					break
-				}
-			}
-			mbLen := len(mb)
-			if mbLen == 0 {
+			stmts, _, err := c.parser.Parse(string(e.Query), "", "")
+			if err != nil {
+				log.Errorf("parse query err %v", err)
 				continue
 			}
-
-			// the first last is table name, the second last is database name(if exists)
-			if len(mb[mbLen-2]) == 0 {
-				db = e.Schema
-			} else {
-				db = mb[mbLen-2]
-			}
-			table = mb[mbLen-1]
-
-			savePos = true
-			force = true
-			c.ClearTableCache(db, table)
-			log.Infof("table structure changed, clear table cache: %s.%s\n", db, table)
-			if err = c.eventHandler.OnTableChanged(string(db), string(table)); err != nil && errors.Cause(err) != schema.ErrTableNotExist {
-				return errors.Trace(err)
-			}
-
-			// Now we only handle Table Changed DDL, maybe we will support more later.
-			if err = c.eventHandler.OnDDL(pos, e); err != nil {
-				return errors.Trace(err)
+			for _, stmt := range stmts {
+				switch t := stmt.(type) {
+				case *ast.RenameTableStmt:
+					for _, tableInfo := range t.TableToTables {
+						db := tableInfo.OldTable.Schema.String()
+						table := tableInfo.OldTable.Name.String()
+						c.updateTable(db, table, e, pos, ev.Header.Timestamp)
+					}
+				case *ast.AlterTableStmt:
+					db := t.Table.Schema.String()
+					table := t.Table.Name.String()
+					c.updateTable(db, table, e, pos, ev.Header.Timestamp)
+				case *ast.DropTableStmt:
+					for _, table := range t.Tables {
+						db := table.Schema.String()
+						table := table.Name.String()
+						c.updateTable(db, table, e, pos, ev.Header.Timestamp)
+					}
+				case *ast.CreateTableStmt:
+					db := t.Table.Schema.String()
+					table := t.Table.Name.String()
+					c.updateTable(db, table, e, pos, ev.Header.Timestamp)
+				case *ast.TruncateTableStmt:
+					db := t.Table.Schema.String()
+					table := t.Table.Name.String()
+					c.updateTable(db, table, e, pos, ev.Header.Timestamp)
+				}
```

Comment:
I think we can keep 

```
savePos = true
force = true
```
and save code in L176 ~ L180

Reply:
agree! and maybe we can use a `default` case to do the `continue`.

Reply:
if we save code in L176~L180, we may lost  some stmt case. we get stmts  from parser and do a range for each stmt, if we keep this two vars true in` QueryEvent` case.there may be too  much save action

Reply:
let's sort out it, one `query event` contains one `query statement(DDL)` and one `position`, and one `query statement(DDL)` contains multiple ddl operations, right?  And we should do `save position` and only once for every `query event`,  but in `for stmts` range, we would do it many times, right?